### DISCUSSION TITLE
How to implement Swiper v9 whit Ionic 📦

### DISCUSSION THREAD

**Discussion by yigsvnsla** [nolimits4web/swiper#6384] [Show and tell]
# How to implement Swiper v9 whit Ionic 📦
##### this post was written with the purpose of helping new people looking to get started with the package or planning to update its dependencies, but being so recent... there seems to be no published solutions on the subject 😄, happy coding...
------------
![](https://i.imgur.com/NTeeBSX.jpeg)

------------
Anteriormente este paquete tenia un soporte para un modulo de componentes que se integraba muy facilmente con **Ionic**/**Angular** ... en esta version dan el salto a los **Web Component** nativos, que como tal no es un problema para **Angular,** pero si rompe en parte la esctrutura de instalacion y uso del mismo teniendo que recurir a el uso del `CUSTOM_ELEMENTS_SCHEMA ` y puede que algunos desarrolladores les pueda incomodar dejar expuesto la encapsulacion de sus componentes, pero a otros les da una ventaja con la llegada de los ***"Stand Alone Components"*** de **Angular ***v15*...

[stackoverflow.com/questions/75326426/how-to-use-swiper-9-with-angular](https://stackoverflow.com/a/75413910/14594701 "stackoverflow.com/questions/75326426/how-to-use-swiper-9-with-angular")

> "Llevo 2 dias buscando informacion desde el momento de esta publicacion, y la actualizacion lleva alrededor de 9 a 10 dias de liberada la release"

Y eh visto algunos ejemplo de implementacion con **Angular ** nativo y otros con **Ionic/Angular** el cual estos tienen cierto parecido en sus respuestas, consisten en retrasar el rendizado del `<swiper-container> `usando `setTimeout()`
> https://github.com/ionic-team/ionic-framework/issues/26722#issuecomment-1422581682
>
> @muuvmuuv
> I managed to initialize it with a timeout:
>
> ```
>   ngAfterViewInit(): void {
>         Object.assign(this.swiperElement.nativeElement, this.config);
>         setTimeout(() => {
>             this.swiperElement.nativeElement.initialize();
>         })
>     }
> ```

Esto representa para mi una mala practica o quizas para otros una solucion eficaz, una de las respuestas en el repositiorio de el equipo de **Ionic** se da a entender que si, existe un retraso en la forma que **Swiper** se monta y desmonta sus componentes con el uso de `<ion-router outlet>` puesto que el mismo es una envoltura del `<router-outlet>` nativo de **Angular.**...
>[issues/26722#](https://github.com/ionic-team/ionic-framework/issues/26722#issuecomment-1422629598 "issues/26722#")
> The behavior of `ion-router-outlet` and `router-outlet` differ slightly which may be the reason why it works with `router-outlet`. I still recommend filing an issue with the Swiper team. I am not familiar with the internals of Swiper 9, so the Swiper team will have a better idea of how Swiper mounts and unmounts.

Otros usuarios reportan el hecho de que al momento de  instanciar **Swiper** este no inicia de manera correcta debido a el retrazo ya mencionado y luego esta el caso que desde **Angular** al tener de referencia un type `Element` o `HTMLElement` tal cual lo dice la [documentacion](https://swiperjs.com/element#parameters-as-props "documentacion") de **Swiper** no podremos acceder a el metodo `intialize()`  ya que este metodo viene de una clase que se encuenta en fichero `swiper/element/swiper-elements.js,` si bien desde javascript vanilla se puede acceder a metodos o propiedades sin tipar (de manera dinamica).

> **Expected Behavior**
nativeElement.initialize

>**Will Trow**
ERROR Error: Uncaught (in promise): TypeError: mod is not a function
TypeError: mod is not a function

**Steps to Reproduce**
```html
<swiper-container  [keyboard]="true"  [pagination]="true"  #swiper  [init]="false" >
	<swiper-slide> ...  </swiper-slide>
</swiper-container>
```
```typescript
@ViewChild('swiper', { static: false }) swiper;

ngAfterViewInit(): void {
	const params = {
		modules: [
			Keyboard, 
			Pagination, 
			Navigation, 
			IonicSwiper
		]
	};
	Object.assign(this.swiper.nativeElement, params);
    this.swiper.nativeElement.initialize(); // this will throw error
}
```

Al momento de tener la referencia desde un `@ViewChild()` desde el dom, este nos dice que esa propiedad no existe y por ende recurrir a ignorar ciertas lineas usando `//@ts-ignore` o convirtiendo la referencia en ``any`` `this.swiper?.nativeElement as any).initialize()`

```javascript
    // @ts-ignore
    this.swiper?.nativeElement.initialize()
```

```javascript
  (  this.swiper?.nativeElement as any ).initialize()
```

sin mencionar que a el momento de importar el type `SwiperComponent `desde el fichero de tipos _`swiper-elements.d.ts`_  nos da un tipado de HTMLElemnt cuando deberia tener tambien los tipos de de la clase `swiper-element` que estan en `swiper-element.js`...

 ![](https://media.discordapp.net/attachments/466773781814181889/1073688698346012823/image.png)

![](https://media.discordapp.net/attachments/466773781814181889/1073688698551537755/image.png)

![](https://cdn.discordapp.com/attachments/466773781814181889/1073688698752876644/image.png)

![](https://cdn.discordapp.com/attachments/466773781814181889/1073688699012907109/image.png)

------------

Bueno quizas es mucho texto... pero si es algo incomodo ya que en discord y una que otra comunidad veo que siguen en la v8 del paquete y no ven luz con este cambio, pero quieren unirse a el progreso... mostrare a continuacion cual fue el metodo que me funciono para poder instaciar **Swiper** en mi proyecto **ionic** *v6 * con **Angular**  *v15.1.2*

![](https://media.tenor.com/AGgVj_aylZUAAAAC/coraline-dad.gif)
*hora del codigo...*

Utilizando los enfoques antes mencionados y conceptos usados, en mi humilde opinion creo que esta es la opcion mas mas sujeta a como estamos acostumbrados a usar con **Angular**

```html
<!-- Creamos nuestro template tal cual el ejemplo de la documentacion de SwiperCreamos nuestro template tal cual el ejemplo de la documentacion de Swiper -->
<ion-content>
  <ion-grid>
    <ion-row>
      <ion-col>
        <swiper-container 
			#swiper 
			init="false" 
			navigation="true" 
			pagination="true" 
			scrollbar="true" >
          <swiper-slide>Slide 1</swiper-slide>
          <swiper-slide>Slide 2</swiper-slide>
          <swiper-slide>Slide 3</swiper-slide>
        </swiper-container>
      </ion-col>
    </ion-row>
  </ion-grid>
</ion-content>
```
```scss
swiper-container {
  width: 90%;
  height: 100%;
}

swiper-slide {
  text-align: center;
  font-size: 18px;
  background: lightgray;
  height: 5rem;
  width: 5rem;
  display: flex;
  justify-content: center;
  align-items: center;
}

swiper-slide img {
  display: block;
  width: 100%;
  height: 100%;
  object-fit: cover;
}
```

```javascript
  @ViewChild("swiper", { static: false }) swiper?: ElementRef<{ swiper: Swiper }>
  private swiperConfig?: SwiperOptions
  constructor(
    private modalController: ModalController,
  ) { }

  ngOnInit() {
    register()
    this.swiperConfig = {
      on: {  init() { }  }
    }
    if (this.swiper?.nativeElement) {
      Object.assign(this.swiper?.nativeElement, this.swiperConfig);
    }
  }

  ngAfterViewInit() {
    console.log("ngAfterViewInit", this.swiper?.nativeElement.swiper);
  }

  ionViewDidEnter() {
    console.log("ionViewDidEnter", this.swiper?.nativeElement.swiper);
  }

  ionViewWillEnter() {
    console.log("ionViewWillEnter", this.swiper?.nativeElement.swiper);
  }
```
debido a el retraso a causa de montar y desmontar el componente `<swiper-component>` vemos en consola que este se instacia en el ultimo gancho de vida de **Ionic** que es  `ionViewDidEnter()`, ya que este se dispara cuando todo el componente esta en cargado en pantalla.

![image](https://user-images.githubusercontent.com/55502763/218245174-fe0b5e44-9453-470a-84c4-b680c0ed7020.png)

Quizas te estaras preguntando: *¿ que hacer en el caso de usar **Angular** sin **Ionic** ?*

podriamos usar el live hook  `ngAfterViewChecked()` esperando el chequeo de angular para iniciar el componente... pero este solo se limita a retornar `null` y sin registrar nuestro **Swiper**.

![image](https://user-images.githubusercontent.com/55502763/218245151-323a11ef-9077-43d5-91c9-96f9e9bb9c75.png)

podriamos usar el evento `ngAfterViewInit()`  para porder registrar nuestro **Swiper**,
la unica limitante es que este no tiene ningun tipo de retraso y parece ser de momento la unica opcion el usar la funcion `setTimeout()` para poder instanciar **Swiper**
```javascript
  ngAfterViewInit() {
    setTimeout(() => console.log("ngAfterViewInit", this.swiper?.nativeElement.swiper));
  }
```

![image](https://user-images.githubusercontent.com/55502763/218245241-88ac71f5-8dbc-4f0b-a1e8-9b0d9584fabe.png)

lo gracioso es ver que nuestro componente sin el **Swiper**, pero si en consola XD.

solo faltaria inicializar el **Swiper** engañando a typeScript usando los "trucos" mencionados unos parrafos mas arriba

```javascript
// no Ionic  
ngAfterViewInit() {
	// @ts-ignore
	setTimeout(() => this.swiper?.nativeElement.initialize());
}

//  Ionic
ionViewDidEnter() {
	// @ts-ignore
	this.swiper?.nativeElement.initialize()
}
```
me parece mas atractivo el `//@ts-ignore` que usar `any` ya que para algunos les parecera mas natura el intepretar que hay algo raro ahi y es mejor ignorarlo.

![](https://cdn.discordapp.com/attachments/1032010359332155444/1073855170745745470/2023-02-11_01-35-00.gif)

y ya eso es todo... solo queria extender este post con el tema sobre el `StyleInyection` en el registro del **Swiper** y ver que nos prepara el futuro
```javascript
register(injectStyles?:boolean)
```

------------

![](https://cdn.discordapp.com/attachments/696497370195820575/1073848375818063902/image.png)

**Comment by nolimits4web** [comment]
Translated with www.DeepL.com/Translator (free version)


# How to implement Swiper v9 whit Ionic 📦
#### During this recent update of the swiper package, one of the major concerns of the community is the implementation of the package in Angular and its use together with Ionic or other component libraries.

##### I would love to write this post in English... but my ability with the language is very weak, and I prefer to do it with my native language 😉....

##### this post was written with the purpose of helping new people looking to get started with the package or planning to update its dependencies, but being so recent... there seems to be no published solutions on the subject 😄, happy coding...
------------
![](https://i.imgur.com/NTeeBSX.jpeg)

------------
Previously this package had a support for a component module that integrated very easily with **Ionic**/**Angular** ... in this version they make the jump to the native **Web Component**, which as such is not a problem for **Angular,** but it does break part of the installation structure and use of the same having to resort to the use of the `CUSTOM_ELEMENTS_SCHEMA` and some developers may be uncomfortable to leave exposed the encapsulation of their components, but to others it gives them an advantage with the arrival of the ***"Stand Alone Components "*** of **Angular ***v15*...

[stackoverflow.com/questions/75326426/how-to-use-swiper-9-with-angular](https://stackoverflow.com/a/75413910/14594701 "stackoverflow.com/questions/75326426/how-to-use-swiper-9-with-angular")

> "I've been looking for information for 2 days since the time of this post, and the update is about 9 to 10 days after the release."

And I have seen some implementation examples with native **Angular ** and others with **Ionic/Angular** which have some similarity in their answers, they consist in delaying the `<swiper-container> rendering `using `setTimeout()`.
> https://github.com/ionic-team/ionic-framework/issues/26722#issuecomment-1422581682
>
> @muuvmuuv
> I managed to initialize it with a timeout:
>
> ```
> ngAfterViewInit(): void {
> Object.assign(this.swiperElement.nativeElement, this.config);
> setTimeout(() => {
> this.swiperElement.nativeElement.initialize();
> })
> }
> ```

This represents for me a bad practice or maybe for others an effective solution, one of the answers in the **Ionic** team repository implies that yes, there is a delay in the way **Swiper** mounts and unmounts its components with the use of `<ion-router outlet>` since it is a wrapper for the native `<router-outlet>` of **Angular.**....
>[issues/26722#](https://github.com/ionic-team/ionic-framework/issues/26722#issuecomment-1422629598 "issues/26722#")
> The behavior of `ion-router-outlet` and `router-outlet` differ slightly which may be the reason why it works with `router-outlet`. I still recommend filing an issue with the Swiper team. I am not familiar with the internals of Swiper 9, so the Swiper team will have a better idea of how Swiper mounts and unmounts.

Other users report the fact that when instantiating **Swiper** it does not start correctly due to the aforementioned delay and then there is the case that from **Angular** when referencing an `Element` or `HTMLElement` type as stated in the [documentation](https://swiperjs. com/element#parameters-as-props "documentacion") of **Swiper** we will not be able to access the `intialize()` method since this method comes from a class that is in file `swiper/element/swiper-elements.js,` although from javascript vanilla you can access to methods or properties without typing (in a dynamic way).

> **Expected Behavior**
nativeElement.initialize

> **Will Trow**
ERROR Error: Uncaught (in promise): TypeError: mod is not a function
TypeError: mod is not a function

**Steps to Reproduce**
```html
<swiper-container [keyboard]="true" [pagination]="true" #swiper [init]="false" >
	<swiper-slide> ...  </swiper-slide>
</swiper-container>
```

```typescript
@ViewChild('swiper', { static: false }) swiper;

ngAfterViewInit(): void {
	const params = {
		modules: [
			Keyboard, 
			Pagination, 
			Navigation, 
			IonicSwiper
		]
	};
	Object.assign(this.swiper.nativeElement, params);
    this.swiper.nativeElement.initialize(); // this will throw error
}
```

At the moment of having the reference from an `@ViewChild()` from the dom, it tells us that this property does not exist and so we resort to ignoring certain lines using `//@ts-ignore` or converting the reference to ``any`` `this.swiper?.nativeElement as any).initialize()`.

```javascript
    // @ts-ignore
    this.swiper?.nativeElement.initialize()
```

```javascript
  ( this.swiper?.nativeElement as any ).initialize()
```


not to mention that when importing the `SwiperComponent` type from the _`swiper-elements.d.ts`_ type file it gives us a HTMLElemnt type when it should also have the types of the `swiper-element` class that are in `swiper-element.js`....

 ![](https://media.discordapp.net/attachments/466773781814181889/1073688698346012823/image.png)

![](https://media.discordapp.net/attachments/466773781814181889/1073688698551537755/image.png)

![](https://cdn.discordapp.com/attachments/466773781814181889/1073688698752876644/image.png)

![](https://cdn.discordapp.com/attachments/466773781814181889/1073688699012907109/image.png)

------------

Well maybe it is a lot of text... but if it is something uncomfortable since in discord and some other community I see that they are still in the v8 package and do not see light with this change, but they want to join the progress... I will show below what was the method that worked for me to install **Swiper** in my project **ionic** *v6 *with **Angular** *v15.1.2*.

![](https://media.tenor.com/AGgVj_aylZUAAAAC/coraline-dad.gif)
*time of the code...*

Using the above mentioned approaches and concepts used, in my humble opinion I think this is the most subject option as we are used to use with **Angular**.

```html
<!-- We create our template as the example in the Swiper documentationWe create our template as the example in the Swiper documentation -->
<ion-content>
  <ion-grid>
    <ion-row>
      <ion-col>
        <swiper-container 
			#swiper 
			init="false" 
			navigation="true" 
			pagination="true" 
			scrollbar="true" >
          <swiper-slide>Slide 1</swiper-slide>
          <swiper-slide>Slide 2</swiper-slide>
          <swiper-slide>Slide 3</swiper-slide>
        </swiper-container>
      </ion-col>
    </ion-row>
  </ion-grid>
</ion-content>
```
```scss
swiper-container {
  width: 90%;
  height: 100%;
}

swiper-slide {
  text-align: center;
  font-size: 18px;
  background: lightgray;
  height: 5rem;
  width: 5rem;
  display: flex;
  justify-content: center;
  align-items: center;
}

swiper-slide img {
  display: block;
  width: 100%;
  height: 100%;
  object-fit: cover;
}
```

```javascript
  @ViewChild("swiper", { static: false }) swiper?: ElementRef<{ swiper: Swiper }>
  private swiperConfig?: SwiperOptions
  constructor(
    private modalController: ModalController,
  ) { }

  ngOnInit() {
    register()
    this.swiperConfig = {
      on: { init() { } }
    }
    if (this.swiper?.nativeElement) {
      Object.assign(this.swiper?.nativeElement, this.swiperConfig);
    }
  }

  ngAfterViewInit() {
    console.log("ngAfterViewInit", this.swiper?.nativeElement.swiper);
  }

  ionViewDidEnter() {
    console.log("ionViewDidEnter", this.swiper?.nativeElement.swiper);
  }

  ionViewWillEnter() {
    console.log("ionViewWillEnter", this.swiper?.nativeElement.swiper);
  }
```
due to the delay caused by mounting and unmounting the `<swiper-component>` component we see in console that this is triggered in the last **Ionic** life hook which is `ionViewDidEnter()`, since this is triggered when the whole component is loaded on screen.

![image](https://user-images.githubusercontent.com/55502763/218245174-fe0b5e44-9453-470a-84c4-b680c0ed7020.png)

You may be wondering: *what to do in the case of using **Angular** without **Ionic** ?

we could use the live hook `ngAfterViewChecked()` waiting for the angular check to start the component... but it just returns `null` and without registering our **Swiper**.

![image](https://user-images.githubusercontent.com/55502763/218245151-323a11ef-9077-43d5-91c9-96f9e9bb9c75.png)

we could use the `ngAfterViewInit()` event to register our **Swiper**,
the only limitation is that this does not have any type of delay and it seems to be for the moment the only option to use the `setTimeout()` function to be able to instantiate **Swiper**.
```javascript
  ngAfterViewInit() {
    setTimeout((() => console.log("ngAfterViewInit", this.swiper?.nativeElement.swiper));
  }
```

![image](https://user-images.githubusercontent.com/55502763/218245241-88ac71f5-8dbc-4f0b-a1e8-9b0d9584fabe.png)

the funny thing is to see that our component without the **Swiper**, but if in console XD.

The only thing left to do is to initialize the **Swiper** by fooling typeScript using the "tricks" mentioned a few paragraphs above

```javascript
// no Ionic  
ngAfterViewInit() {
	// @ts-ignore
	setTimeout(() => this.swiper?.nativeElement.initialize());
}

// Ionic
ionViewDidEnter() {
	// @ts-ignore
	this.swiper?.nativeElement.initialize()
}
```
I find `//@ts-ignore` more attractive than `any` as some people will find it more natural to interpret that there is something weird there and it is better to ignore it.


![](https://cdn.discordapp.com/attachments/1032010359332155444/1073855170745745470/2023-02-11_01-35-00.gif)

and that's all... I just wanted to extend this post with the topic about `StyleInyection` in the **Swiper** registry and see what the future holds for us.
```javascript
register(injectStyles?:boolean)
```

------------

![](https://cdn.discordapp.com/attachments/696497370195820575/1073848375818063902/image.png)

**Comment by yigsvnsla** [comment]
**Post Update: Important bug in method past to swiper instances**

because angular and the `**CUSTOM_ELEMENTS_SCHEMA**` are not instantiated in the **Angular** rendering tree through lazy loading, a slight bug is noticed when mounting and unmounting Swiper Elements that amusingly causes a slight advance before initializing the Swiper Web Component... I'll be elaborating a post using the Swiper core natively that seems to work much better in performance and in coupling to the native angular live hooks.

![image](https://user-images.githubusercontent.com/55502763/218628761-ba0b3fca-121c-4eb3-a354-56abac9fcf04.png)

this weird effect is more noticeable in nested routes

It seems that when registering several swipers using the `register()` function, they are scoped to the same instance even though you have the registries in different modules

happy coding 😊

**Reply by muuvmuuv** [reply]
We really need to figure out where this problem is coming from, I still think its Ionic and the ion-router-outlet but it could also be some wrong init process on Swiper side. Angular's router works fine.

**Reply by yigsvnsla** [reply]
in effect... the problem is how swiper mounts and unmounts the Web Component, and I add to that... the Angular router does not index in its tree, elements based on custom elements

It's a mistake on both sides in my opinion.

**Comment by malua** [comment]
It seems that with **Angular 19**, **Ionic 8** and **Swiper 11** this not exactly correct anymore. 

It works now by putting the init call inside `ngAfterViewInit` wrapped with a `setTimeout` with `0ms`. I have a custom swiper component that wraps the swiper custom element.
```
swiperRef = viewChild<ElementRef>('swiperContainer');

ngAfterViewInit() {
    setTimeout(() => {
      this.swiperRef().nativeElement.initialize();
    }, 0);
  }
```

This works for routed pages (`ion-router-outet`) as well as in Ionic modals.
Calling init in `ionViewDidEnter` is not necessary anymore and it didn't work for modals anyway, as `ionViewDidEnter` is not called there.
